File: Sem 5/CS/asgn9/sandbox.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <seccomp.h>
#include <sys/wait.h>
#include <signal.h>
#include <string.h>

// Define the folder where file operations are allowed
#define ALLOWED_FOLDER "/home/vbuntu/Desktop/Vatsal/college/Sem 5/CS/asgn9/sandbox/"
// Set a timeout for program execution in seconds
#define TIMEOUT_SECONDS 5
// Maximum heap size allowed (10 MB)
#define MAX_HEAP_SIZE (10 * 1024 * 1024) // 10 MB
// Maximum length for file paths
#define PATH_MAX 100000

// Signal handler for timeout
void timeout_handler(int signum) {
    printf("Execution Timeout. Terminating sandbox.\n");
    exit(1);
}

// Function to restrict heap size
void restrict_heap_size() {
    struct rlimit limit;
    // Set soft and hard limits for heap allocation to 10 MB
    limit.rlim_cur = MAX_HEAP_SIZE; // Soft limit
    limit.rlim_max = MAX_HEAP_SIZE; // Hard limit

    // Apply the limit using setrlimit system call
    if (setrlimit(RLIMIT_AS, &limit) != 0) {
        perror("Failed to restrict heap size");
        exit(1);
    }
}

// Function to check if the file path is inside the allowed folder
int is_path_allowed(const char *path) {
    char real_path[PATH_MAX];
    // Resolve the real absolute path of the file, handling any symbolic links
    realpath(path, real_path);

    // Compare the beginning of the resolved path with the allowed folder path
    if (strncmp(real_path, ALLOWED_FOLDER, strlen(ALLOWED_FOLDER)) == 0) {
        return 1;
    }
    return 0; 
}

void write_to_file(const char *folder, const char *filename, const char *data) {
    char filepath[PATH_MAX];
    // Construct the full file path by combining folder and filename
    snprintf(filepath, sizeof(filepath), "%s/%s", folder, filename);

    // Check if the constructed file path is allowed
    if (!is_path_allowed(filepath)) {
        fprintf(stderr, "Permission Denied: Write operation outside of sandbox folder!\n");
        exit(1);
    }

    // Open the file for writing
    FILE *file = fopen(filepath, "w");
    if (file == NULL) {
        // If opening the file fails, print an error message and exit
        perror("Failed to open file");
        exit(1);
    }
    
    fprintf(file, "%s", data);
    fclose(file);
    printf("File written successfully to %s\n", filepath);
}

int main() {
    // Set up the timeout by associating SIGALRM with the timeout handler function
    signal(SIGALRM, timeout_handler);
    // Start the countdown for the timeout (5 seconds)
    alarm(TIMEOUT_SECONDS);

    restrict_heap_size();

    // Sandbox task: Write to a file in the allowed folder
    write_to_file(ALLOWED_FOLDER, "sandboxed_file.txt", "Data inside the sandbox");

    // Simulate a long-running task that exceeds the timeout
    sleep(10); // This will be interrupted by the timeout after 5 seconds

    return 0;
}

File: Sem 5/CS/asgn9/sandbox.py
import threading
import time
import os

# Restrict File Access to Specific Directory
ALLOWED_FOLDER = os.path.abspath("/home/vbuntu/Desktop/Vatsal/college/Sem 5/CS/asgn9/sandbox")  # Ensures full path

# Function to write data to a file only within the allowed folder
def write_to_file(folder, filename, data):
    # Check if the allowed folder exists
    if not os.path.exists(ALLOWED_FOLDER):
        raise FileNotFoundError(f"The allowed folder {ALLOWED_FOLDER} does not exist.")

    folder_path = os.path.abspath(folder)  # Get absolute path for folder
    if folder_path.startswith(ALLOWED_FOLDER):
        file_path = os.path.join(folder, filename)
        with open(file_path, 'w') as f:
            f.write(data)
        print(f"File written successfully to {file_path}")
    else:
        raise PermissionError("Write operation outside of sandbox folder is prohibited!")

# Function to handle sandboxed code execution
def sandbox_code():
    try:
        # Write to file in the allowed sandbox folder
        write_to_file(ALLOWED_FOLDER, "sandboxed_file.txt", "Data inside the sandbox")
        time.sleep(10)  # Simulate a long-running task
    except Exception as e:
        print(f"Error: {e}")

# Class to handle the sandbox timeout using threading
class SandboxThread(threading.Thread):
    def __init__(self, timeout):
        super().__init__()  # Proper initialization of the Thread
        self.timeout = timeout
        self.result = None

    def run(self):
        try:
            self.result = sandbox_code()
        except Exception as e:
            self.result = e

# Function to run the sandbox code with a timeout
def run_with_timeout(timeout_seconds):
    thread = SandboxThread(timeout=timeout_seconds)
    thread.start()
    thread.join(timeout_seconds)

    if thread.is_alive():
        print("Execution Timeout. Terminating sandbox.")
        # Terminate the thread (simulated since thread termination is complex in Python)
        raise Exception("Sandbox execution timed out")
    else:
        print("Sandbox executed successfully")

# Main execution
if __name__ == "__main__":
    try:
        # Run the sandbox code with a 5-second timeout
        run_with_timeout(5)
    except Exception as e:
        print(e)
